### PR TITLE
Add BlockLimitReachedEvent to StateCheckpoint

### PR DESCRIPTION
**Pull request by igor-aptos**
Cleanup so BlockExecutor always marks StateCheckpoint as non-retry. simplifies a lot of logic outside.

In the StateCheckpoint, if block was ended early, we add BlockLimitReachedEvent, with information of why block limit was reached.

TODO:
 - use it in gas estimation, for whether blocks were full or not
 - gate adding event by field in onchain config

### CODE REVIEW

**Comment by igor-aptos** [comment]
Current dependencies on/for this PR:
* main
  * **PR #10879** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/10879?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #11110** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11110?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #10943** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/10943?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #11167** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11167?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11167?utm_source=stack-comment).

File: types/src/account_config/events/new_block.rs [hunk 1]
```diff
@@ -141,3 +141,12 @@ impl MoveStructType for BlockResource {
 }
 
 impl MoveResource for BlockResource {}
+
+/// (keep this doc in sync with the `struct BlockLimitReachedEvent` in Move.)
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct BlockLimitReachedEvent {
```
Comment (danielxiangzl) [inline types/src/account_config/events/new_block.rs:147]: Should we also emit a fee receipt (exe / io gas cost + output size) when hitting the limit, or even for every block. Maybe current / future gas mechanism can leverage this information?

Comment (igor-aptos) [reply types/src/account_config/events/new_block.rs:147]: not sure we want to add events to non-StateCheckpoint transaction. it is going to be very odd for user transaction that ends a block to have an event emitted (that it wasn't charged for)

maybe I can add to non-epoch-ending blocks

but, will add more details to this event irrespectively

Comment (danielxiangzl) [reply types/src/account_config/events/new_block.rs:147]: Depending on whether such information is useful for the gas mechanism, but I guess those two flags are sufficient already.

File: aptos-move/block-executor/src/executor.rs [hunk 2]
```diff
@@ -690,7 +701,7 @@ where
                     approx_output_size,
                 );
 
-                if accumulator.should_end_block_parallel() {
+                if txn_idx == scheduler.num_txns() - 1 && accumulator.should_end_block_parallel() {
```
Comment (danielxiangzl) [inline aptos-move/block-executor/src/executor.rs:704]: `txn_idx < scheduler.num_txns() - 1`?

File: aptos-move/block-executor/src/executor.rs [hunk 3]
```diff
@@ -690,7 +701,7 @@ where
...
                     // Set the execution output status to be SkipRest, to skip the rest of the txns.
                     last_input_output.update_to_skip_rest(txn_idx);
```
Comment (danielxiangzl) [inline aptos-move/block-executor/src/executor.rs:706]: `txn_idx` is the txn that exceeds limit? Shouldn't we set `txn_idx+1` to skip?

File: aptos-move/block-executor/src/executor.rs (hunk 2 above)
Comment (igor-aptos) [reply aptos-move/block-executor/src/executor.rs:704]: argh yes!

File: aptos-move/block-executor/src/executor.rs (hunk 3 above)
Comment (igor-aptos) [reply aptos-move/block-executor/src/executor.rs:706]: I think SkipRest means skip after this one

**Review by danielxiangzl — COMMENTED** [review]
Thanks for saving my ugly state checkpoint changes!
IIUC, after this the state checkpoint will always be at the end of the block, even when the block gets cut? If yes, does is this fine to storage? cc @msmouse

**Comment by igor-aptos** [comment]
It should be same as before - if block is ended by a transaction - we don't add StateCheckpoint. if it is added by limit, we just move where we add the block. so should be same for storage (unless I have a bug)

File: aptos-move/block-executor/src/executor.rs (hunk 3 above)
Comment (danielxiangzl) [reply aptos-move/block-executor/src/executor.rs:706]: Ah correct

**Comment by danielxiangzl** [comment]
I think this needs to be changed accordingly as well, otherwise (1) duplicated state checkpoint will be added when the block is not cut, and (2) state checkpoint will not be in `to_keep` when block is cut. The variant we need to maintain here is `to_keep` always ends with state checkpoint or reconfig.
https://github.com/aptos-labs/aptos-core/blob/fd15c8721b7289759084ce77cbe9e149acf6189c/execution/executor/src/components/apply_chunk_output.rs#L193

Also running e2e test to see if the change works?

**Comment by msmouse** [comment]
hmm.. I realized replay and state sync won't work without hacks? In that sense this should be part of the StateCheckpoint Txn, not in the event..

**Comment by igor-aptos** [comment]
moved out the cleanup of StateCheckpoint into #11253, and leaving the rest for here.

**Pull request closed without merging**